src/hathor.c

```diff
@@ -92,64 +68,111 @@ void compress_public_key(unsigned char *value) {
+    // prepend version
+    out[0] = P2PKH_VERSION_BYTE;
+    os_memmove(out+1, hash, 20);
+    // sha256d of above and get first 4 bytes (checksum)
+    sha256d(out, 21, checksum_buffer);
+    os_memmove(out+21, checksum_buffer, 4);
+}
 
-    if (inlen > B58_MAX_INPUT_SIZE) {
-        return -1;
-    }
+void pubkey_to_address(cx_ecfp_public_key_t *public_key, uint8_t *out) {
+    unsigned char hash_buffer[20];
+    // get compressed pubkey
+    compress_public_key(public_key->W);
+    // get hash160
+    hash160(public_key->W, 33, hash_buffer);
+    pubkey_hash_to_address(hash_buffer, out);
+}
 
-    while ((zeroCount < inlen) && (in[zeroCount] == 0)) {
-        ++zeroCount;
+/**
+ * Validates that a script has the format of P2PKH. Throws an exception if doesn't.
+ * P2PKH scripts have the format:
+ *   [OP_DUP, OP_HASH160, pubkey_hash_len, pubkey_hash, OP_EQUALVERIFY, OP_CHECKSIG]
+ * Considering that pubkey hashes have 20 bytes and the values of the opcodes:
+ *   [0x76, 0xA9, 20, pubkey_hash, 0x88, 0xAC]
+ */
+void validate_p2pkh_script(uint8_t *in) {
+    uint8_t p2pkh[] = {0x76, 0xA9, 20, 0x88, 0xAC};
+    if (os_memcmp(p2pkh, in, 3) != 0 || os_memcmp(p2pkh + 3, in + 23, 2) !=0) {
+        THROW(SW_INVALID_PARAM);
     }
+}
 
-    outputSize = (inlen - zeroCount) * 138 / 100 + 1;
-    stopAt = outputSize - 1;
-    for (startAt = zeroCount; startAt < inlen; startAt++) {
-        int carry = in[startAt];
-        for (j = outputSize - 1; (int)j >= 0; j--) {
-            carry += 256 * buffer[j];
-            buffer[j] = carry % 58;
-            carry /= 58;
-
-            if (j <= stopAt - 1 && carry == 0) {
-                break;
-            }
-        }
-        stopAt = j;
+/*
+ * Parses the output as either a 4 or 8-byte unsigned integer.
+ *
+ * Returns the position in buffer after parsing the value.
+ */
+uint8_t* parse_output_value(uint8_t *in, size_t inlen, uint64_t *value) {
+    uint8_t *buf = in;
+    uint64_t tmp = 0;
+    // if first bit is 1, it's 8 bytes long. Otherwise, it's 4
+    bool flag = ((0x80 & in[0]) ? true : false);
+    if (flag) {
+        assert_length(11, inlen);    // value + token_data + script_len
+        tmp = U8BE(in, 0);
+        tmp = (-1)*tmp;
+        buf += 8;
+    } else {
+        tmp = U4BE(in, 0);
+        buf += 4;
     }
+    os_memcpy(value, &tmp, 8);
+    return buf;
+}
 
-    j = 0;
-    while (j < outputSize && buffer[j] == 0) {
-        j += 1;
-    }
+/**
+ * Parses a tx output from the input data. Returns a pointer to the end of parsed data.
+ */
+uint8_t* parse_output(uint8_t *in, size_t inlen, tx_output_t *output) {
+    uint8_t *buf = in;
+    assert_length(7, inlen);    // value + token_data + script_len
+    buf = parse_output_value(buf, inlen, &output->value);
+    output->token_data = *buf;
+    buf++;
+    uint16_t script_len = U2BE(buf, 0);
+    buf += 2;
+    assert_length(script_len, inlen - 7);
+    //XXX considering only p2pkh, without timelock
```

Comment:
Maybe move this comment to the `validate_p2pkh_script` method.

Reply:
Done.